server/src/main/java/org/opensearch/common/remote/AbstractRemoteWritableBlobEntity.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.remote;
+
+import org.opensearch.common.blobstore.BlobPath;
+import org.opensearch.core.compress.Compressor;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.gateway.remote.ClusterMetadataManifest.UploadedMetadata;
+
+import static org.opensearch.gateway.remote.RemoteClusterStateUtils.PATH_DELIMITER;
+
+/**
+ * An extension of {@link RemoteWriteableEntity} class which caters to the use case of writing to and reading from a blob storage
+ *
+ * @param <T> The class type which can be uploaded to or downloaded from a blob storage.
+ */
+public abstract class AbstractRemoteWritableBlobEntity<T> implements RemoteWriteableEntity<T> {
+
+    protected String blobFileName;
+
+    protected String blobName;
+    private final String clusterUUID;
+    private final Compressor compressor;
+    private final NamedXContentRegistry namedXContentRegistry;
+    private String[] pathTokens;
+
+    public AbstractRemoteWritableBlobEntity(
+        final String clusterUUID,
+        final Compressor compressor,
+        final NamedXContentRegistry namedXContentRegistry
+    ) {
+        this.clusterUUID = clusterUUID;
+        this.compressor = compressor;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    public abstract BlobPathParameters getBlobPathParameters();
+
+    public String getFullBlobName() {
+        return blobName;
+    }
+
+    public String getBlobFileName() {
+        if (blobFileName == null) {
+            String[] pathTokens = getBlobPathTokens();
+            if (pathTokens == null || pathTokens.length < 1) {
+                return null;
+            }
+            blobFileName = pathTokens[pathTokens.length - 1];
+        }
+        return blobFileName;
+    }
+
+    public String[] getBlobPathTokens() {
+        if (pathTokens != null) {
+            return pathTokens;
+        }
+        if (blobName == null) {
+            return null;
+        }
+        pathTokens = blobName.split(PATH_DELIMITER);
+        return pathTokens;
+    }
+
+    public abstract String generateBlobFileName();
+
+    public String clusterUUID() {
+        return clusterUUID;
+    }
+
+    public abstract UploadedMetadata getUploadedMetadata();
+
+    public void setFullBlobName(BlobPath blobPath) {
```

Comment:
why do we have 2 methods one generateBlobFileName and another set? Also can we call generateBlobFileName in the constructor itself to ensure it is called only once.

Reply:
The two methods are `setFullBlobName` and `generateBlobFileName` are different.
Full blob name includes the path. For example `/path/filename` is the full blob name and `filename` is the blob file name. Blob file name can contain the timestamp, so `generateBlobFileName` is invoked just before upload to keep the timestamp accurate.
FullBlobName is set by the RemoteBlobStore implementation as it has the knowledge of the base blob path.